src/classes/BDI_ObjectMatchingLogic.cls

```diff
@@ -0,0 +1,156 @@
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Batch Data Import
+* @group-content ../../ApexDocContent/BatchDataImport.htm
+* @description Virtual class containing the standard logic for matching source and target records.  
+* Can be overriden by clients who need custom matching logic.
+*/
+
+public with sharing virtual class BDI_ObjectMatchingLogic {
+
+    public virtual BDI_ObjectWrapper[] matchObjects(BDI_ObjectWrapper[] objWraps, BDI_DataImportService dataImportService) {
+        if (objWraps != null && objWraps.size() > 0 && objWraps[0].fieldMappings != null) {
+
+            Boolean shouldMatch = false;
+
+            //Note it is assumed that all object wraps passed into this class will be using the same
+            //field mappings.
+            Data_Import_Field_Mapping__mdt[] fieldMappings = objWraps[0].fieldMappings;
+
+            String targetObject = objWraps[0].objMapping.Object_API_Name__c;
+            String importedRecordStatusField = objWraps[0].objMapping.Imported_Record_Status_Field_Name__c;
+            String importedRecordField = objWraps[0].objMapping.Imported_Record_Field_Name__c;
+
+            String matchSourceField;
+            String matchTargetField;
+            Integer matchSourceCount = 0;
+
+            //Determine what the source field is for matching
+            for (Data_Import_Field_Mapping__mdt fieldMapping : fieldMappings) {
+                if (fieldMapping.Match_On_This_Mapping__c == 'Yes') {
+                    matchSourceField = fieldMapping.Source_Field_API_Name__c;
+                    matchTargetField = fieldMapping.Target_Field_API_Name__c;
+
+                    matchSourceCount++;
+                }
+            }
+
+            Map<String,BDI_ObjectWrapper[]> objWrapsByMatchingSourceValue = new Map<String,BDI_ObjectWrapper[]>();
+
+            //For basic matching only one match field is allowed, if there is more than one then
+            //add an error to all data import records;
+            if (matchSourceCount > 1) {
+                for (BDI_ObjectWrapper objWrap : objWraps) {             
+                    objWrap.logError(dataImportService.statusDIFailed(),System.label.bdiObjMatchErrorTooManyMatchingFields);
+                }
+            } else if (matchSourceCount == 1) {
+
+                // If the source field for the match on the DI has a value then we will attempt to match.
+                // also add the source strings for matching to a list.
+                for (BDI_ObjectWrapper objWrap : objWraps) {
+
+                    String sourceValue = (String)objWrap.dataImport.get(matchSourceField);
```

Comment:
I can't remember, are we restricting type of the matching field at all? Is there any reason that casting to string here would be a problem depending on the type of the field?

Reply:
On the UI side I was planning on limiting it to be datatypes that could be converted to string easily, and were not long text (ie Text, Lookup, Id).  We can expand or change the options as needed, but Ali was ok with these options for now.